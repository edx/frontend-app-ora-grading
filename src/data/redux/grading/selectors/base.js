import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import * as module from './base';

export const rootSelector = ({ grading }) => grading;
export const mkSimpleSelector = (key) => createSelector(
  [module.rootSelector],
  (grading) => grading[key],
);
export const rootKeys = StrictDict({
  selected: 'selected',
  activeIndex: 'activeIndex',
  current: 'current',
  gradeData: 'gradeData',
  gradingData: 'gradingData',
});

export const simpleSelectors = {
  selected: mkSimpleSelector(rootKeys.selected),
  activeIndex: mkSimpleSelector(rootKeys.activeIndex),
  current: mkSimpleSelector(rootKeys.current),
  gradeData: mkSimpleSelector(rootKeys.gradeData),
  gradingData: mkSimpleSelector(rootKeys.gradingData),
};

/**
 * returns the length of the list of selected submissions
 * @return {number} selected submission list length
 */
export const selectionLength = createSelector(
  [module.simpleSelectors.selected],
  (selected) => selected.length,
);

export default {
  ...simpleSelectors,
  selectionLength,
};
