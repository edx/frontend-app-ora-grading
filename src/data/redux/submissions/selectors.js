import _ from 'lodash';
import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import { lockStatuses } from 'data/services/lms/constants';

export const simpleSelectors = {
  allSubmissions: state => state.submissions.allSubmissions,
};

/**
 * Returns the submission list in default order for the table.
 */
export const listData = createSelector(
  [simpleSelectors.allSubmissions],
  (allSubmissions) => {
    const submissionIds = Object.keys(allSubmissions);
    const submissionList = submissionIds.map(id => {
      const { gradeStatus, lockStatus, ...rest } = allSubmissions[id];
      const gradingStatus = (lockStatus === lockStatuses.unlocked ? gradeStatus : lockStatus);
      return { gradingStatus, ...rest };
    });
    return _.sortBy(
      submissionList,
      ['submissionDate'],
    );
  },
);

export default StrictDict({
  ...simpleSelectors,
  listData,
});
