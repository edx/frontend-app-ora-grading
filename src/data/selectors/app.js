import { StrictDict } from 'utils';

export const simpleSelectors = {
  showReview: state => state.app.showReview,
  showRubric: state => state.app.showRubric,
  grading: state => state.app.grading,
  courseMetadata: state => state.app.courseMetadata,
  oraName: state => state.app.oraMetadata.name,
  oraPrompt: state => state.app.oraMetadata.prompt,
  oraTypes: state => state.app.oraMetadata.type,
};

export default StrictDict({
  ...simpleSelectors,
});
