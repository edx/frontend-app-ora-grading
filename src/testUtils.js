// eslint-disable-next-line import/prefer-default-export
export const formatMessage = (msg, values) => {
  let message = msg.defaultMessage;
  if (values === undefined) { return message; }
  Object.keys(values).forEach(key => {
    // eslint-disable-next-line
    message = message.replace(`{${key}}`, values[key]);
  });
  return message;
};
