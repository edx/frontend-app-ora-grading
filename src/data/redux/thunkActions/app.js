import { StrictDict } from 'utils';

import { actions } from 'data/redux';
import { locationId } from 'data/constants/app';
import { initializeApp } from './requests';

/**
 * initialize the app, loading ora and course metadata from the api, and loading the initial
 * submission list data.
 */
export const initialize = () => (dispatch) => {
  dispatch(initializeApp({
    locationId,
    onSuccess: (response) => {
      dispatch(actions.app.loadOraMetadata(response.oraMetadata));
      dispatch(actions.app.loadCourseMetadata(response.courseMetadata));
      dispatch(actions.submissions.loadList(response.submissions));
    },
  }));
};

/**
 * initialize the app, loading ora and course metadata from the api, and loading the initial
 * submission list data.
 */
export const reloadSubmissions = () => (dispatch) => {
  dispatch(initializeApp({
    locationId,
    reload: true,
    onSuccess: (response) => {
      dispatch(actions.submissions.loadList(response.submissions));
    },
  }));
};

export default StrictDict({ initialize, reloadSubmissions });
