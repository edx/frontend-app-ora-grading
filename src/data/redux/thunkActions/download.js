import JSZip from 'jszip';
import FileSaver from 'file-saver';

import { RequestKeys } from 'data/constants/requests';
import { selectors } from 'data/redux';

import { networkRequest } from './requests';
import * as module from './download';

/**
 * Generate a manifest file content based on files object
 * @param {obj[]} files - list of file entries with downloadUrl, name, description, and size
 * @return {string} - manifest text file content.
 */
export const genManifest = (files) => files.map(
  (file) => `Filename: ${file.name}\nDescription: ${file.description}\nSize: ${file.size}`,
).join('\n\n');

/**
 * Returns the zip filename
 * @return {string} - zip download file name
 */
export const zipFileName = () => {
  const currentDate = new Date().getTime();
  return `ora-files-download-${currentDate}.zip`;
};

/**
 * Zip the blob output of a set of files with a manifest file.
 * @param {obj[]} files - list of file entries with downloadUrl, name, and description
 * @param {blob[]} blobs - file content blobs
 * @return {Promise} - zip async process promise.
 */
export const zipFiles = (files, blobs) => {
  const zip = new JSZip();
  zip.file('manifest.txt', module.genManifest(files));
  blobs.forEach((blob, i) => zip.file(files[i].name, blob));
  return zip.generateAsync({ type: 'blob' }).then(
    zipFile => FileSaver.saveAs(zipFile, module.zipFileName()),
  );
};

/**
 * Download a file and return its blob is successful, or null if not.
 * @param {obj} file - file entry with downloadUrl
 * @return {blob} - file blob or null
 */
export const downloadFile = (file) => fetch(file.downloadUrl).then(resp => (
  resp.ok ? resp.blob() : null
));

/**
 * Download blobs given file objects.  Returns a promise map.
 * @param {obj[]} files - list of file entries with downloadUrl, name, and description
 * @return {Promise[]} - Promise map of download attempts (null for failed fetches)
 */
export const downloadBlobs = (files) => Promise.all(files.map(module.downloadFile));

/**
 * Download all files for the selected submission as a zip file.
 * Throw error and do not download zip if any of the files fail to fetch.
 */
export const downloadFiles = () => (dispatch, getState) => {
  const { files } = selectors.grading.selected.response(getState());
  dispatch(networkRequest({
    requestKey: RequestKeys.downloadFiles,
    promise: module.downloadBlobs(files).then(blobs => {
      if (blobs.some(blob => blob === null)) {
        throw Error('Fetch Failed');
      }
      module.zipFiles(files, blobs);
    }),
  }));
};

export default {
  downloadFiles,
};
