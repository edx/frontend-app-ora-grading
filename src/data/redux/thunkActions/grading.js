import { StrictDict } from 'utils';

import { RequestKeys } from 'data/constants/requests';
import { actions, selectors } from 'data/redux';

import * as module from './grading';
import requests from './requests';

/**
 * Prefetch the "next" submission in the selected queue.  Only fetches the response info.
 */
export const prefetchNext = () => (dispatch, getState) => {
  dispatch(requests.fetchSubmissionResponse({
    requestKey: RequestKeys.prefetchNext,
    submissionId: selectors.grading.next.submissionId(getState()),
    onSuccess: (response) => {
      dispatch(actions.grading.preloadNext(response));
    },
  }));
};

/**
 * Prefetch the "previous" submission in the selected queue.  Only fetches the response info.
 */
export const prefetchPrev = () => (dispatch, getState) => {
  dispatch(requests.fetchSubmissionResponse({
    requestKey: RequestKeys.prefetchPrev,
    submissionId: selectors.grading.prev.submissionId(getState()),
    onSuccess: (response) => {
      dispatch(actions.grading.preloadPrev(response));
    },
  }));
};

/**
 * Fetch the target neighbor submission's status, start grading if in progress,
 * dispatches load action with the response (injecting submissionId).  If hasNeighbor,
 * also dispatches the prefetchAction to pre-fetch the new neighbor's response.
 * @param {string} submissionId - target submission id
 * @param {action} loadAction - redux action/thunkAction to load the submission status
 * @param {bool} hasNeighbor - is there a new neighbor to be pre-fetched?
 * @param {action} prefetchAction - redux action/thunkAction to prefetch the new
 *   neighbor's response.
 */
export const fetchNeighbor = ({
  submissionId,
  loadAction,
  hasNeighbor,
  prefetchAction,
}) => (dispatch) => {
  dispatch(requests.fetchSubmissionStatus({
    submissionId,
    onSuccess: (response) => {
      dispatch(loadAction({ ...response, submissionId }));
      if (hasNeighbor) { dispatch(prefetchAction()); }
    },
  }));
};

/**
 * Fetches the current status for the "next" submission in the selected queue,
 * and calls loadNext with it to update the current selection index info.
 * If the new index has a next submission available, preload its response.
 */
export const loadNext = () => (dispatch, getState) => {
  dispatch(module.fetchNeighbor({
    loadAction: actions.grading.loadNext,
    hasNeighbor: selectors.grading.next.doesExist(getState()),
    prefetchAction: module.prefetchNext,
    submissionId: selectors.grading.next.submissionId(getState()),
  }));
};

/**
 * Fetches the current status for the "previous" submission in the selected queue,
 * and calls loadPrev with it to update the current selection index info.
 * If the new index has a previous submission available, preload its response.
 */
export const loadPrev = () => (dispatch, getState) => {
  dispatch(module.fetchNeighbor({
    loadAction: actions.grading.loadPrev,
    hasNeighbor: selectors.grading.prev.doesExist(getState()),
    prefetchAction: module.prefetchPrev,
    submissionId: selectors.grading.prev.submissionId(getState()),
  }));
};

/**
 * Load a list of selected submissionIds, sets the app to review mode, and fetches the current
 * selected submission's full data (grade data, status, and rubric).
 * Then loads current selection and prefetches neighbors.
 * @param {string[]} submissionIds - ordered list of submissionIds for selected submissions
 */
export const loadSelectionForReview = (submissionIds) => (dispatch, getState) => {
  dispatch(requests.fetchSubmission({
    submissionId: submissionIds[0],
    onSuccess: (response) => {
      dispatch(actions.grading.updateSelection(submissionIds));
      dispatch(actions.grading.loadSubmission({
        ...response,
        submissionId: submissionIds[0],
      }));
      dispatch(actions.app.setShowReview(true));
      if (selectors.grading.next.doesExist(getState())) {
        dispatch(module.prefetchNext());
      }
      if (selectors.grading.prev.doesExist(getState())) {
        dispatch(module.prefetchPrev());
      }
    },
  }));
};

/**
 * Start grading the current submission.
 * Attempts to lock the submisison, and on a success, sets the local grading state to
 * True, and then loads initializes the grading process with GradeData associated with
 * the current submission.  If there is no grade data, generates an empty grade entry
 * based on the rubric config.
 */
export const startGrading = () => (dispatch, getState) => {
  dispatch(requests.setLock({
    value: true,
    submissionId: selectors.grading.selected.submissionId(getState()),
    onSuccess: (response) => {
      dispatch(actions.app.setShowRubric(true));
      let { gradeData } = response;
      if (!gradeData) {
        gradeData = selectors.app.emptyGrade(getState());
      }
      dispatch(actions.grading.startGrading({ ...response, gradeData }));
    },
  }));
};

/**
 * Cancels the grading process for the current submisison.
 * Releases the lock and dispatches stopGrading on success.
 */
export const cancelGrading = () => (dispatch, getState) => {
  dispatch(requests.setLock({
    value: false,
    submissionId: selectors.grading.selected.submissionId(getState()),
    onSuccess: () => {
      dispatch(module.stopGrading());
    },
  }));
};

/**
 * Stops the grading process for the current submission (local only)
 * Clears the local grade data for the current submission and sets grading state
 * to False
 */
export const stopGrading = () => (dispatch) => {
  dispatch(actions.grading.stopGrading());
};

export const submitGrade = () => (dispatch, getState) => {
  const gradeData = selectors.grading.selected.gradingData(getState());
  const submissionId = selectors.grading.selected.submissionId(getState());
  if (selectors.grading.validation.isValidForSubmit(getState())) {
    dispatch(actions.grading.setShowValidation(false));
    dispatch(requests.submitGrade({
      submissionId,
      gradeData,
      onSuccess: (response) => {
        dispatch(actions.grading.completeGrading(response));
      },
      onFailure: () => {
        // on failure action
      },
    }));
  } else {
    dispatch(actions.grading.setShowValidation(true));
  }
};

export default StrictDict({
  loadSelectionForReview,
  loadNext,
  loadPrev,
  startGrading,
  cancelGrading,
  stopGrading,
  submitGrade,
});
