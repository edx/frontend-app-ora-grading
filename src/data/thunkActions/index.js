import { StrictDict } from 'utils';

import app from './app';
import grading from './grading';

export default StrictDict({
  app,
  grading,
});
