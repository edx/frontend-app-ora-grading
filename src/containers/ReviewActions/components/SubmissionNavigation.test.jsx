import React from 'react';
import { shallow } from 'enzyme';

import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';

import { formatMessage } from 'testUtils';

import {
  SubmissionNavigation,
  mapStateToProps,
  mapDispatchToProps,
} from './SubmissionNavigation';

jest.mock('@edx/paragon', () => ({
  Icon: () => 'Icon',
  IconButton: () => 'IconButton',
}));
jest.mock('@edx/paragon/icons', () => ({
  ChevronLeft: 'ChevronLeft',
  ChevronRight: 'ChevronRight',
}));
jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    grading: {
      prev: {
        doesExist: (state) => ({ prevDoesExist: state }),
      },
      next: {
        doesExist: (state) => ({ nextDoesExist: state }),
      },
      activeIndex: (state) => ({ activeIndex: state }),
      selectionLength: (state) => ({ selectionlength: state }),
    },
  },
}));

describe('SubmissionNavigation component', () => {
  describe('component', () => {
    const props = {
      intl: { formatMessage },
      activeIndex: 4,
      selectionLength: 5,
    };
    beforeEach(() => {
      props.loadNext = jest.fn().mockName('this.props.loadNext');
      props.loadPrev = jest.fn().mockName('this.props.loadPrev');
    });
    test('snapshot: no prev submission (disabled)', () => {
      expect(shallow(
        <SubmissionNavigation {...props} activeIndex={0} hasNextSubmission />,
      )).toMatchSnapshot();
    });
    test('snapshot: no next submission (disabled)', () => {
      expect(shallow(
        <SubmissionNavigation {...props} hasPrevSubmission />,
      )).toMatchSnapshot();
    });
  });
  describe('mapStateToProps', () => {
    let mapped;
    const testState = { some: 'test-state' };
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('activeIndex loads from grading.activeIndex', () => {
      expect(mapped.activeIndex).toEqual(selectors.grading.activeIndex(testState));
    });
    test('hasNextSubmission loads from grading.next.doesExist', () => {
      expect(mapped.hasNextSubmission).toEqual(selectors.grading.next.doesExist(testState));
    });
    test('hasPrevSubmission loads from grading.prev.doesExist', () => {
      expect(mapped.hasPrevSubmission).toEqual(selectors.grading.prev.doesExist(testState));
    });
    test('selectionLength loads from grading.selectionLength', () => {
      expect(mapped.selectionLength).toEqual(selectors.grading.selectionLength(testState));
    });
  });
  describe('mapDispatchToProps', () => {
    it('loads loadNext from thunkActions.grading.loadNext', () => {
      expect(mapDispatchToProps.loadNext).toEqual(thunkActions.grading.loadNext);
    });
    it('loads loadPrev from thunkActions.grading.loadPrev', () => {
      expect(mapDispatchToProps.loadPrev).toEqual(thunkActions.grading.loadPrev);
    });
  });
});
