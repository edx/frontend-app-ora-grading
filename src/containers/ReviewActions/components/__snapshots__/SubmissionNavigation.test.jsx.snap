// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionNavigation component component snapshot: no next submission (disabled) 1`] = `
<span
  className="submission-navigation"
>
  <IconButton
    alt="Load previous submission"
    className="ml-1"
    disabled={false}
    iconAs={[Function]}
    onClick={[MockFunction this.props.loadPrev]}
    size="inline"
    src="ChevronLeft"
  />
  <span
    className="ml-1"
  >
    <FormattedMessage
      defaultMessage="{current} of {total}"
      description="Submission navigation location label"
      id="ora-grading.ReviewActions.navigationLabel"
      values={
        Object {
          "current": 5,
          "total": 5,
        }
      }
    />
  </span>
  <IconButton
    alt="Load next submission"
    className="ml-1"
    disabled={true}
    iconAs={[Function]}
    onClick={[MockFunction this.props.loadNext]}
    size="inline"
    src="ChevronRight"
  />
</span>
`;

exports[`SubmissionNavigation component component snapshot: no prev submission (disabled) 1`] = `
<span
  className="submission-navigation"
>
  <IconButton
    alt="Load previous submission"
    className="ml-1"
    disabled={true}
    iconAs={[Function]}
    onClick={[MockFunction this.props.loadPrev]}
    size="inline"
    src="ChevronLeft"
  />
  <span
    className="ml-1"
  >
    <FormattedMessage
      defaultMessage="{current} of {total}"
      description="Submission navigation location label"
      id="ora-grading.ReviewActions.navigationLabel"
      values={
        Object {
          "current": 1,
          "total": 5,
        }
      }
    />
  </span>
  <IconButton
    alt="Load next submission"
    className="ml-1"
    disabled={false}
    iconAs={[Function]}
    onClick={[MockFunction this.props.loadNext]}
    size="inline"
    src="ChevronRight"
  />
</span>
`;
